File: Aio-Algorithms-And-Solutions/Algorithms/DataStructures/Stack/stack.py
from typing import TypeVar, Generic, List

T = TypeVar('T')

class Stack(Generic[T]):
    def __init__(self, capacity: int = 4):
        self._items: List[T] = []
        self._capacity = capacity

    def push(self, item: T) -> None:
        self._items.append(item)

    def pop(self) -> T:
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self._items.pop()

    def peek(self) -> T:
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self._items[-1]

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def size(self) -> int:
        return len(self._items)

    def clear(self) -> None:
        self._items.clear() 

File: Aio-Algorithms-And-Solutions/Algorithms/DataStructures/Stack/Stack.cs
public class Stack<T>
{
    private T[] items;
    private int top;
    private const int DefaultCapacity = 4;

    public Stack(int capacity = DefaultCapacity)
    {
        items = new T[capacity];
        top = -1;
    }

    public void Push(T item)
    {
        if (top == items.Length - 1)
        {
            Resize();
        }
        items[++top] = item;
    }

    public T Pop()
    {
        if (IsEmpty)
            throw new InvalidOperationException("Stack is empty");
            
        return items[top--];
    }

    public T Peek()
    {
        if (IsEmpty)
            throw new InvalidOperationException("Stack is empty");
            
        return items[top];
    }

    public bool IsEmpty => top == -1;
    
    public int Count => top + 1;

    private void Resize()
    {
        T[] newItems = new T[items.Length * 2];
        Array.Copy(items, newItems, items.Length);
        items = newItems;
    }
} 